trio/_channel.py

```diff
@@ -4,10 +4,11 @@
 import attr
 from outcome import Error, Value
 
-from . import _core
 from .abc import SendChannel, ReceiveChannel
 from ._util import generic_function
 
+import trio.hazmat
```

Comment:
Should just be `import trio`

Reply:
That was my first instinct as well, but when I've tried it locally and with the automated tests on commit, I'm getting an undefined attribute "hazmat" error.  My guess is it has to do with importing the specific part of trio instead of the entire module, but I'm afraid it's not apparent to me at the moment (probably a simple thing I'm overlooking).  

Any ideas?  Should I do something along the lines of "from trio import hazmat as trio_hazmat" for clarity sake?  Thanks!


----

[...]_channel.py", line 125, in MemorySendChannel                                            
    @trio.hazmat.enable_ki_protection
AttributeError: module 'trio' has no attribute 'hazmat'

Reply:
Oh, I see, ick. You're going to have the same problem in `_sync.py` too, I think.

The reason we can mostly get away with importing `trio`, inside modules that are part of the implementation of `trio`, is because in most cases we don't actually access the `trio` namespace until runtime. So, when python runs the module's top-level code, the `trio` namespace isn't fully set up yet... but that's fine, we aren't touching it then anyway. But, a function decorator like `@trio.hazmat.enable_ki_protection` runs at import time, so you do need `trio.hazmat` to be already set at import time.

Unfortunately, doing `import trio.hazmat` is also problematic :-/. I'm pretty sure what's happening is that it forces `trio.hazmat` to be imported early, much earlier than `trio/__init__.py` would do it... which works right now but is super confusing.

I guess the simplest solution would be to make a special exception for `enable_ki_protection`, and access it using `from ._core import enable_ki_protection` (while still using `import trio` for everything else). Maybe:

```python
# Import directly from _core instead of trio.hazmat, to avoid circular import
from ._core import enable_ki_protection
```

Reply:
Ah, after I commented I was wondering if being a decorator had anything to do with it - forgot it runs at import.  That makes good sense to me - I can see importing hazmat early could lead to unexpected behavior in the future (and painful debugging).

Reply:
This actually occurs elsewhere other than decorators.  I'm going ahead and resolving by referencing _core directly instead of bringing in hazmat early.